File: 2020/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for l in s:
            if l in '({[':
                stack.append(l)
            else:
                if len(stack) == 0: return False
                if l == ')' and stack.pop() != '(': return False
                elif l == ']' and stack.pop() != '[': return False
                elif l == '}' and stack.pop() != '{': return False
        return len(stack) == 0


# @lc code=end

sol = Solution()
print(sol.isValid('{[]}'))
print(sol.isValid('([)]'))

File: js/20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  let matches = {
    '(': ')',
    '[': ']',
    '{': '}',
  }
  let stack = []
  for (let i=0; i<s.length; i++) {
    if (stack.length > 0 && matches[stack[stack.length-1]] === s[i])
      stack.pop()
    else
      stack.push(s[i])
  }
  return stack.length === 0
};
// @lc code=end

